src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,259 @@
+      error_code::invalid_replica_assignment,
+      "Duplicate replica ids in partition reassignment replica list",
+      [](const reassignable_partition& partition) {
+          auto replicas_copy = *partition.replicas;
+          std::sort(replicas_copy.begin(), replicas_copy.end());
+          auto no_duplicates_it = std::adjacent_find(
+            replicas_copy.cbegin(), replicas_copy.cend());
+          return no_duplicates_it == replicas_copy.cend();
+      });
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Invalid broker id in replica list",
+      [](const reassignable_partition& partition) {
+          auto negative_node_id_it = std::find_if(
+            partition.replicas->begin(),
+            partition.replicas->end(),
+            [](const model::node_id& node_id) { return node_id < 0; });
+          return negative_node_id_it == partition.replicas->end();
+      });
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Replica assignment has brokers that are not alive",
+      [&all_node_ids](const reassignable_partition& partition) {
+          auto unkown_broker_id_it = std::find_if(
+            partition.replicas->begin(),
+            partition.replicas->end(),
+            [all_node_ids](const model::node_id& node_id) {
+                return std::find(
+                         all_node_ids.begin(), all_node_ids.end(), node_id)
+                       == all_node_ids.end();
+            });
+          return unkown_broker_id_it == partition.replicas->end();
+      });
+
+    // Store any invalid partitions in the response
+    topic_response.partitions = std::move(invalid_partitions);
+    resp_it = std::move(topic_response);
+
+    return valid_partitions_end;
+}
+
+template<>
+ss::future<response_ptr> alter_partition_reassignments_handler::handle(
+  request_context ctx, [[maybe_unused]] ss::smp_service_group ssg) {
+    alter_partition_reassignments_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    log_request(ctx.header(), request);
+    alter_partition_reassignments_response resp;
+
+    if (!ctx.authorized(
+          security::acl_operation::alter, security::default_cluster_name)) {
+        resp.data.error_code = error_code::cluster_authorization_failed;
+        resp.data.error_message = ss::sstring{
+          error_code_to_str(error_code::cluster_authorization_failed)};
+        co_return co_await ctx.respond(std::move(resp));
+    }
+
+    // TODO(@NyaliaLui) - Check if we are the controller
+
+    resp.data.responses.reserve(request.data.topics.size());
+    auto all_node_ids = ctx.metadata_cache().node_ids();
+
+    for (reassignable_topic& topic : request.data.topics) {
```

Comment:
nit: I see you don't like `auto` how come?

Reply:
@NyaliaLui i assume when you say that auto makes it more difficult to debug you mean that you'd like to know what the type has been deduced to? for this case I do something like

```
auto x = <something i'm not sure what type it is>;
std::string (or int, ...) = x;
```

and then the clangd diagnostics in your IDE should present a message with what type X is.

Reply:
Oo that's a good idea. Thanks!